api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -895,6 +895,10 @@ message HttpConnectionManager {
   // This should be set to ``false`` in cases where Envoy's view of the downstream address may not correspond to the
   // actual client address, for example, if there's another proxy in front of the Envoy.
   google.protobuf.BoolValue add_proxy_protocol_connection_state = 53;
+
+  // Whether the HCM will clear hop by hop response headers. The default value is true, and it is only false by default when using
```

Comment:
I don't think a good API has this conditional default. Is it possible to have a consistent default (true) and then just use the boolean to configure the difference? For Envoy Mobile, the environment could program this presumably?

Reply:
I think another feasible way is to explicitly state that Envoy Mobile will ignore this parameter and always retain hop-by-hop response headers, so there won't be any inconsistency in default values. Otherwise, it will result in compatibility issues.

Reply:
If this is only for Envoy mobile, I don't think we need an API for it.  I think we can implement this purely in C++ 
We create the API listener in internal_engine.cc and can set the boolean there and avoid APIs entirely

Reply:
Ah sorry just read in the issue that there's other use cases than E-M.   In that case would it be better to have it be part of UCN now that we're moving towards having a consistent way of sanitizing headers?  cc @yanavlasov

Reply:
Also (sorry to comment bomb) I think it's going to have to be more than just a boolean "clear or not clear"
for E-M simply not clearing would work. For the proxy case, I think we'd have to consider combinations and permutations of protocols - I think there's some hop by hop headers which are say valid for upstream HTTP/1.1 but not valid for downstream HTTP/2 so we'd have to conditionally clear them in the codec based on support. I _think_ we didn't bother because we unconditionally clear.  If we're going to go the permanent config route I'd hope we'd put this with other header sanitation variables and mark it hidden until we think we've sorted out all those corner cases because I suspect it's going to cause at least a few problems

Reply:
@alyssawilk Thank you for providing the information. I realize that it is inaccurate to solve this issue by simply not clearing the hop-by-hop headers. In fact, we need to set the Connection and Keep-Alive headers for the envoy proxy layer in order to adapt to certain special clients, instead of passing through these headers from upstream. However, currently there is no distinction made when mutating response headers whether these headers were added through HCM config or returned by upstream, which prevents us from setting them. Do you have any suggestions on how to resolve this problem?
By the way, I would also like to understand what you mean by "UCN" mentioned above.

Reply:
A reasonable solution that I have considered is to refer to the [keepalive_timeout](https://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout) directive in Nginx. Set the [idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-msg-config-core-v3-httpprotocoloptions) to the Keep-Alive Header when handling HTTP/1, but do not set it for HTTP/2 or HTTP/3.

Reply:
@alyssawilk Hello, I have solved this issue in a new way, similar to how Nginx does it. Could you please review it again? Thank you.